### DISCUSSION TITLE
Issue with submitAnswer overwriting data for requests from the same client IP

### DISCUSSION THREAD

**Discussion by Chem-Stitch** [ANLAB-KAIST/KENSv3#139] [Project Q&A 1. Application Layer (Socket)] [answered]
Hello,

I'm currently working on a server implementation where I face an issue with handling multiple requests from the same client IP. The core of the problem lies in the submitAnswer function call within my server code.

Problem Encountered: When multiple requests arrive from the same client IP, the submitAnswer call seems to overwrite the previous log data with the latest request's data. This is problematic as I need to retain the data for each request independently, regardless of whether they come from the same IP.

Expected Behavior: Each call to submitAnswer should independently log the request data for every request.
Actual Behavior: If multiple requests come from the same client IP, only the latest request's data is retained, and previous data is overwritten.

Questions:
Is there a recommended approach or best practice within this framework to ensure that data for each request is logged independently, even if the requests come from the same client IP?

I appreciate any insights or suggestions you can provide to help resolve this issue. Thank you in advance for your assistance!

**Comment by Pusnow** [comment]
Which test cases do you fail? PA1 only includes test cases that each client connects its own server only once. 

> Questions:
> Is there a recommended approach or best practice within this framework to ensure that data for each request is logged independently, even if the requests come from the same client IP?

Currently, the submitAnswer function uses IP->Answer `std::map`, and it handles very simple test scenarios only, so you cannot log all requests come from the same client IP

In the future, as you suggested, we can extend KENS to support more complicated scenario using more complicated data structures like IP->Vector<Answer>.

Thank you for your suggestion.

**Reply by Chem-Stitch** [reply]
I failed every '-all' cases and 'ManytoMany' case. This is the test case that I made for test.

TEST_F(EchoTesting, OnetoManyAllSpecific) {
  SetUp({"server-hello9"}, 4,
        {
            {0, 0, "echo-hello1"},    {0, 0, "whoru"},
            {0, 0, "whoami"},         {0, 0, "hello"}
        });
  runTest();
  checkTest(true);
}

In this case, doesn't the same client (with the same ID) connect to one server multiple times and deliver different content each time? In this case, content seems to be overwritten when server run submitAnswer.

**Reply by Pusnow** [reply, marked as answer]
For test cases similar to yours, we skip checking server-side's answer by passing `true` to `void checkTest(bool skip_server)`.

If you are failing to the test cases, you probably submit wrong client-side answers.

**Reply by Chem-Stitch** [reply]
I downloaded the latest version of KENS again from github and now it works. Sorry for the confusion. Thank you!
